src/SamplesApp/SamplesApp.UITests/TestFramework/ImageAssert.cs

```diff
@@ -341,136 +292,136 @@ public static void HasPixels(FileInfo screenshot, params ExpectedPixels[] expect
-						if (expectedColor.IsEmpty)
-						{
-							continue;
-						}
-
-						var pixelX = expectation.X + col + offsetX;
-						var pixelY = expectation.Y + lin + offsetY;
-						var pixel = bitmap.GetPixel(pixelX, pixelY);
-
-						var expected = ToArgbCode(expectedColor);
-						var actual = ToArgbCode(pixel);
-
-						if (!AreSameColor(expectedColor, pixel, expectation.ColorTolerance))
-						{
-							isSuccess = false;
-							result.AppendLine($"{col},{lin}: [{pixelX},{pixelY}] (expected: {expected} | actual: {actual})");
-						}
-					}
-
-				if (failMessage == default) // so we keep only for offset 0,0
-				{
-					failMessage = result.ToString();
-				}
+				case LocationToleranceKind.PerRange:
+					isSuccess = GetLocationOffsets(expectation)
+						.Any(offset => GetPixelCoordinates(expectation)
+							.All(pixel => ValidatePixel(actualBitmap, expectation, expectedToActualScale, pixel, offset, report)));
+					break;
+
+				case LocationToleranceKind.PerPixel:
+					isSuccess = GetPixelCoordinates(expectation)
+						.All(pixel => GetLocationOffsets(expectation)
+							.Any(offset => ValidatePixel(actualBitmap, expectation, expectedToActualScale, pixel, offset, report)));
+					break;
+
+				default:
+					throw new ArgumentOutOfRangeException(nameof(expectation.Tolerance.OffsetKind));
+			}
 
-				return isSuccess;
+			if (isSuccess) // otherwise the report has already been full-filled
+			{
+				report?.AppendLine("\tOK");
 			}
+			return isSuccess;
 		}
-		#endregion
 
-		private static int GetColorDistance(Color a, Color b)
+		private static IEnumerable<(int x, int y)> GetLocationOffsets(ExpectedPixels expectation)
 		{
-			return
-				Abs(a.A - b.A) +
-				Abs(a.R - b.R) +
-				Abs(a.G - b.G) +
-				Abs(a.B - b.B);
+			for (var offsetX = 0; offsetX <= expectation.Tolerance.Offset.x; offsetX++)
+			for (var offsetY = 0; offsetY <= expectation.Tolerance.Offset.y; offsetY++)
+			{
+				yield return (offsetX, offsetY);
+				if (offsetX > 0) yield return (-offsetX, offsetY);
+				if (offsetY > 0) yield return (offsetX, -offsetY);
+				if (offsetX > 0 && offsetY > 0) yield return (-offsetX, -offsetY);
+			}
 		}
 
-		private static bool AreSameColor(Color a, Color b, byte tolerance = 0)
+		private static IEnumerable<Point> GetPixelCoordinates(ExpectedPixels expectation)
 		{
-			// comparing ARGB values, because 'named colors' are not considered equal to their unnamed equivalents(!)
-			return
-				Abs(a.A - b.A) <= tolerance &&
-				Abs(a.R - b.R) <= tolerance &&
-				Abs(a.G - b.G) <= tolerance &&
-				Abs(a.B - b.B) <= tolerance;
-		}
+			var stepX = (int)Math.Min(1, expectation.Tolerance.DiscreteValidation.x);
+			var stepY = (int)Math.Min(1, expectation.Tolerance.DiscreteValidation.y);
 
-		private static string ToArgbCode(Color color)
-			=> $"{color.A:X2}{color.R:X2}{color.G:X2}{color.B:X2}";
-	}
+			for (var lin = 0; lin < expectation.Values.GetLength(0); lin+= stepY)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'for' block.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5363624)